cli/compiler.rs

```diff
@@ -1,57 +1,105 @@
-  pub maybe_source_map: Option<Vec<u8>>,
+/// Optional tuple which represents the state of the compiler
+/// configuration where the first is canonical name for the configuration file
+/// and a vector of the bytes of the contents of the configuration file.
+type CompilerConfig = Option<(String, Vec<u8>)>;
+
+/// Information associated with compiled file in cache.
+/// Includes source code path and state hash.
+/// version_hash is used to validate versions of the file
+/// and could be used to remove stale file in cache.
+pub struct CompiledFileMetadata {
+  pub source_path: String,
+  pub version_hash: String,
 }
 
-impl ModuleMetaData {
-  pub fn has_output_code_and_source_map(&self) -> bool {
-    self.maybe_output_code.is_some() && self.maybe_source_map.is_some()
+static SOURCE_PATH: &'static str = "source_path";
+static VERSION_HASH: &'static str = "version_hash";
+
+impl CompiledFileMetadata {
+  // TODO: use serde for deserialization
+  /// Get compiled file metadata from meta_path.
+  /// If operation failed or metadata file is corrupted,
+  /// return None.
+  pub fn load<P: AsRef<Path>>(meta_path: P) -> Option<CompiledFileMetadata> {
+    let meta_path = meta_path.as_ref();
+    let maybe_metadata_string = fs::read_to_string(meta_path).ok();
+    maybe_metadata_string.as_ref()?;
+    CompiledFileMetadata::from_json_string(maybe_metadata_string.unwrap())
   }
 
-  pub fn js_source(&self) -> String {
-    if self.media_type == msg::MediaType::Json {
-      return format!(
-        "export default {};",
-        str::from_utf8(&self.source_code).unwrap()
-      );
-    }
-    match self.maybe_output_code {
-      None => str::from_utf8(&self.source_code).unwrap().to_string(),
-      Some(ref output_code) => str::from_utf8(output_code).unwrap().to_string(),
+  pub fn from_json_string(metadata_string: String) -> Option<Self> {
+    let maybe_metadata_json: serde_json::Result<serde_json::Value> =
+      serde_json::from_str(&metadata_string);
+
+    if let Ok(metadata_json) = maybe_metadata_json {
+      let source_path = metadata_json[SOURCE_PATH].as_str().map(String::from);
+      let version_hash = metadata_json[VERSION_HASH].as_str().map(String::from);
+
+      if source_path.is_none() || version_hash.is_none() {
+        return None;
+      }
+
+      return Some(CompiledFileMetadata {
+        source_path: source_path.unwrap(),
+        version_hash: version_hash.unwrap(),
+      });
     }
+
+    None
   }
-}
 
-type CompilerConfig = Option<(String, Vec<u8>)>;
+  pub fn to_json_string(self: &Self) -> Result<String, serde_json::Error> {
+    let mut value_map = serde_json::map::Map::new();
+
+    value_map.insert(SOURCE_PATH.to_owned(), json!(&self.source_path));
+    value_map.insert(VERSION_HASH.to_string(), json!(&self.version_hash));
+    serde_json::to_string(&value_map)
+  }
+
+  /// Save compiled file metadata to meta_path.
+  pub fn save<P: AsRef<Path>>(self: &Self, meta_path: P) {
```

Comment:
I'm probably missing something - but I don't see where this function is called.. ?

Reply:
Not anymore, it was from original implementation by Kevin